core/history.go

```diff
@@ -0,0 +1,78 @@
+package core
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+)
+
+type History struct {
+	txn db.Transaction
+}
+
+func NewHistory(txn db.Transaction) *History {
+	return &History{txn: txn}
+}
+
+func (h *History) logOldValue(key, value []byte, height uint64) error {
+	return h.txn.Set(binary.BigEndian.AppendUint64(key, height), value)
+}
+
+func (h *History) valueAt(key []byte, height uint64) ([]byte, error) {
+	it, err := h.txn.NewIterator()
+	if err != nil {
+		return nil, err
+	}
+
+	for it.Seek(binary.BigEndian.AppendUint64(key, height)); it.Valid(); it.Next() {
+		seekedKey := it.Key()
+		// seekedKey size should be `len(key) + sizeof(uint64)` and seekedKey should match key prefix
+		if len(seekedKey) != len(key)+8 || !bytes.HasPrefix(seekedKey, key) {
+			break
+		}
+
+		seekedHeight := binary.BigEndian.Uint64(seekedKey[len(key):])
+		if seekedHeight < height {
```

Comment:
Can't imagine a case where this could happen since our implementation only seeks forward `SeekGE`

Reply:
On another note, I think it's good to check for it regardless

Reply:
Yeah, I think you are right. Let's keep it for the sake of being more expressive, it might help down the line when updating the implementation.